crates/network-scanner-net/src/socket.rs

```diff
@@ -203,15 +219,32 @@ impl<'a> Future for ConnectFuture<'a> {
     type Output = std::io::Result<()>;
 
     fn poll(self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
-        let events = self.runtime.check_event_with_id(self.id, true);
-        if events.iter().any(|e| e.is_connect_failed().unwrap_or(false)) {
-            tracing::warn!(?self.socket, ?self.addr, "connect failed");
-            return std::task::Poll::Ready(Err(std::io::Error::new(std::io::ErrorKind::Other, "connect failed")));
-        };
+        let events = self.runtime.check_event_with_id(self.id);
+        for event in events {
+            tracing::trace!(?event, "event found");
+            if event
+                .is_connect_failed() // For linux, failed connection is ERR and HUP, a sigle HUP does not indicate a failed connection
+                .expect("your platform does not support connect failed")
+            {
+                return std::task::Poll::Ready(Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "connection failed",
+                )));
+            }
 
-        if !events.is_empty() {
-            tracing::trace!(?events, "connect success");
-            return std::task::Poll::Ready(Ok(()));
+            // This is a special case, this happens when using epoll to wait for a unconnected TCP socket.
+            // We clearly needs to call connect function, so we break the loop and call connect.
+            #[cfg(target_os = "linux")]
```

Comment:
Linux and Windows have different behavior, Linux shots event to you even before you call connect, we need to identify if the event is "Connect failed" or "Connect successful" or something else (out and hup). Polling does not expose plateform dependent args, so we have to make a banch specific to our needs here. Will disguss with Polling's owner later.